File: codes/python/inheritance/multiple/multiple1.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

class Person(object):
    def __init__(self):
        print("I am a person")

class Machine(object):
    def __init__(self):
        print("I am a machine")

class Android(Machine, Person):
    def __init__(self):
        Machine.__init__(self)
        Person.__init__(self)
        print("So, I am an android")

class Cyborg(Person, Machine):
    def __init__(self):
        Person.__init__(self)
        Machine.__init__(self)
        print("I am a cyborg")


p = Person()
print("=====================")
m = Machine()
print("=====================")
a = Android()
print("=====================")
c = Cyborg()
print("=====================")

File: codes/ruby/inheritance/multiple/multiple1.rb
#!/usr/bin/ruby -w

module PersonMixin
    def info; puts "I am a person"; end
end

module MachineMixin
    def info; puts "I am a machine"; end
end

module InitMixin
    def initialize()
        info2 = method(:info).super_method
        if info2 != nil then info2.call end
        info
        moreInfo
    end
    def moreInfo; end
end

class Person include PersonMixin, InitMixin end

class Machine include MachineMixin, InitMixin end

class Android < Machine
    include PersonMixin, InitMixin
    def moreInfo; puts "So, I am an android"; end
end

class Cyborg < Person
    include MachineMixin, InitMixin
    def moreInfo; puts "So, I am an cyborg"; end
end

# ====== main =======
# ===================

p = Person.new
puts "====================="
m = Machine.new
puts "====================="
a = Android.new
puts "====================="
c = Cyborg.new
puts "====================="
